rfd/00XX-agent-census.md

```diff
@@ -0,0 +1,271 @@
+If this succeeds, the agent is running on Kubernetes.
+
+Afterwards, we'll try to detect in which cloud provider the pod is running on.
+For this, we'll call `client.ServerVersion()` and check if the returned `gitVersion` contains a certain substring specific for each provider:
+
+- in EKS, the git version looks like `"v1.24.8-eks-ffeb93d"` (we'll search for substring `"-eks"`)
+- in GPC ([docs](https://cloud.google.com/kubernetes-engine/docs/release-notes)), the git version looks like `"1.23.14-gke.1800"` (we'll search for substring `"-gke"`)
+
+In AKS, the git version looks like `"v1.25.2"`, so it's not possible to detect this environment using this method. (This is also a problem for Helm charts, as reported in [Azure/AKS#3375](https://github.com/Azure/AKS/issues/3375).)
+
+In the end, `UpstreamInventoryAgentMetadata.ContainerOrchestrator` will be set to:
+- `kubernetes-eks` if on EKS
+- `kubernetes-gke` if on GKE
+- `kubernetes-unknown` otherwise (AKS, other cloud provider, or no cloud provider)
+
+##### 10. Cloud environment
+
+The only way to determine this seems to be by hitting certain HTTP endpoints specific to each cloud environment:
+- AWS ([docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html)): http://169.254.169.254/latest/
+- GCP ([docs](https://cloud.google.com/compute/docs/metadata/overview#parts-of-a-request)): http://metadata.google.internal/computeMetadata/v1/
+- Azure ([docs](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux#access-azure-instance-metadata-service)): http://169.254.169.254/metadata/instance?api-version=2021-02-01
+
+`UpstreamInventoryAgentMetadata.CloudEnvironment` will be set to:
+- `aws` if on AWS
+- `gcp` if on GCP
+- `azure` if on Azure
+
+#### Development plan
+
+The above work will be divided in the following tasks:
+
+1. Add new message type `AgentMetadataEvent` to PreHog.
+2. Add new message type `UpstreamInventoryAgentMetadata`. These will be empty initially, except for the fields common with `UpstreamInventoryHello`. This is okay since being empty means that the field could not be determined, which may happen anyways.
+3. Extend auth server to convert `UpstreamInventoryAgentMetadata` messages to `AgentMetadataEvent` messages and push them to PreHog.
+4. Gradually instrument & add new code that fills each new `UpstreamInventoryAgentMetadata` message field.
+
+We could decide to do step 4 together with step 2 if we don't want to risk adding fields to `UpstreamInventoryAgentMetadata` that possibly won't be used in the end (if for some reason we figure out they can/should not be tracked).
+However, step 4 requires several changes (Go code & files used by the multiple installation methods) which we may want to review separately.
+A similar reasoning also applies to step 1 since each field in `AgentMetadataEvent` should only exist if it also exists in `UpstreamInventoryAgentMetadata`.
+
+### Security
```

Comment:
We pull from a range of environment variables herein, all of which are user controlled data.  Lets assume these are hostile and ensure the design is hardened against that.  Sanitation needs to happen somewhere in infrastructure we control. Perhaps it already does as part of PostHog?  It'd be worth calling out.

Reply:
This will be worth pointing out but it's not going to be a real problem until we start collecting data from OSS; all data that's collected from cloud and enterprise is going to be associated with a license, so if we start noticing weird data, we can exclude the entire source from our analysis (and presumably go ask what's going on).

Reply:
> Security wise: We pull from a range of environment variables herein, all of which are user controlled data. Lets assume these are hostile and ensure the design is hardened against that. Sanitation needs to happen somewhere in infrastructure we control. Perhaps it already does as part of PostHog? It'd be worth calling out.

ac4116a02e8ee1c7318fe5c53938838325a64c59 adds a note about this. @wadells please LMK if the proposed plan sounds good.

Reply:
Thanks  -- that is good enough for me.